mls-rs-uniffi/src/lib.rs

```diff
@@ -0,0 +1,524 @@
+        let key_package = arc_unwrap_or_clone(key_package);
+        let mut group = self.inner().await;
+        let mls_message = group.propose_add(key_package.inner, Vec::new()).await?;
+        Ok(mls_message.into())
+    }
+
+    /// Propose and commit the removal of a member.
+    ///
+    /// The member is representated by the key package in `member`.
+    ///
+    /// See [`mls_rs::group::CommitBuilder::remove_member`] for details.
+    pub async fn remove_member(&self, member: Arc<SigningIdentity>) -> Result<CommitOutput, Error> {
+        let identifier = signing_identity_to_identifier(&member.inner)?;
+        let mut group = self.inner().await;
+        let member = group.member_with_identity(&identifier).await?;
+        let commit_output = group
+            .commit_builder()
+            .remove_member(member.index)?
+            .build()
+            .await?;
+        Ok(commit_output.into())
+    }
+
+    /// Propose to remove a member from this group.
+    ///
+    /// The member is representated by the key package in `member`.
+    /// The result is the welcome message to send to this member.
+    ///
+    /// See [`mls_rs::group::Group::propose_remove`] for details.
+    pub async fn propose_remove_member(
+        &self,
+        key_package_message: Arc<Message>,
+    ) -> Result<Message, Error> {
+        let key_package_message = arc_unwrap_or_clone(key_package_message);
+        let key_package = key_package_message
+            .inner
+            .into_key_package()
+            .ok_or(MlsError::UnexpectedMessageType)?;
+        let identifier = signing_identity_to_identifier(key_package.signing_identity())?;
+        let mut group = self.inner().await;
+        let member = group.member_with_identity(&identifier).await?;
+        let mls_message = group.propose_remove(member.index, Vec::new()).await?;
+        Ok(mls_message.into())
+    }
+
+    /// Encrypt an application message using the current group state.
+    pub async fn encrypt_application_message(&self, message: &[u8]) -> Result<Message, Error> {
+        let mut group = self.inner().await;
+        let mls_message = group
+            .encrypt_application_message(message, Vec::new())
+            .await?;
+        Ok(mls_message.into())
+    }
+
+    /// Process an inbound message for this group.
+    pub async fn process_incoming_message(
+        &self,
+        message: Arc<Message>,
+    ) -> Result<ReceivedMessage, Error> {
+        let message = arc_unwrap_or_clone(message);
+        let mut group = self.inner().await;
+        match group.process_incoming_message(message.inner).await? {
+            group::ReceivedMessage::ApplicationMessage(application_message) => {
+                let sender = Arc::new(
+                    self.index_to_identity(application_message.sender_index)
+                        .await?
+                        .into(),
+                );
+                let data = application_message.authenticated_data;
```

Comment:
This should be `let data = application_message.data().to_vec()`. The `authenticated_data` will usually be empty -- this is sent _unencrypted_ along with the message

Reply:
Thanks! I need to get some unit tests going for this ASAP...